File: app/src/main/java/com/example/firstapp/protocol/ProtocolModels.kt
package com.example.firstapp.protocol

import kotlinx.serialization.Serializable

@Serializable
data class DiscoveryMessage(
    val type: String,
    val deviceId: String,
    val deviceName: String,
    val platform: String,
    val tlsPort: Int? = null,
    val fingerprint: String? = null
)

@Serializable
data class PairRequest(
    val deviceId: String,
    val deviceName: String,
    val publicKey: String,
    val code: String
)

@Serializable
data class PairResponse(
    val deviceId: String,
    val deviceName: String,
    val fingerprint: String,
    val peerPublicKey: String,
    val salt: String,
    val token: String? = null
)

@Serializable
data class SessionRequest(
    val deviceId: String,
    val publicKey: String,
    val token: String? = null
)

@Serializable
data class SessionResponse(
    val peerPublicKey: String,
    val salt: String
)

@Serializable
data class TransferInitRequest(
    val fileName: String,
    val fileSize: Long,
    val mimeType: String?,
    val sha256: String,
    val chunkSize: Int
)

@Serializable
data class TransferInitResponse(
    val transferId: String,
    val accepted: Boolean,
    val missingChunks: List<Int> = emptyList()
)

@Serializable
data class TransferChunkRequest(
    val transferId: String,
    val index: Int,
    val totalChunks: Int,
    val nonce: String,
    val cipherText: String,
    val aad: String
)

@Serializable
data class TransferChunkResponse(
    val received: Boolean
)

@Serializable
data class TextMessageRequest(
    val messageId: String,
    val nonce: String,
    val cipherText: String,
    val aad: String
)

@Serializable
data class TextMessageResponse(
    val received: Boolean
)

File: windows/PulseSend.Core/Protocol/ProtocolModels.cs
using System.Text.Json.Serialization;

namespace PulseSend.Core.Protocol;

public sealed record DiscoveryMessage
{
    public string Type { get; init; } = "";
    public string DeviceId { get; init; } = "";
    public string DeviceName { get; init; } = "";
    public string Platform { get; init; } = "";
    public int? TlsPort { get; init; }
    public string? Fingerprint { get; init; }
}

public sealed record PairRequest
{
    public string DeviceId { get; init; } = "";
    public string DeviceName { get; init; } = "";
    public string PublicKey { get; init; } = "";
    public string Code { get; init; } = "";
}

public sealed record PairResponse
{
    public string DeviceId { get; init; } = "";
    public string DeviceName { get; init; } = "";
    public string Fingerprint { get; init; } = "";
    public string PeerPublicKey { get; init; } = "";
    public string Salt { get; init; } = "";
    public string? Token { get; init; }
}

public sealed record SessionRequest
{
    public string DeviceId { get; init; } = "";
    public string PublicKey { get; init; } = "";
    public string? Token { get; init; }
}

public sealed record SessionResponse
{
    public string PeerPublicKey { get; init; } = "";
    public string Salt { get; init; } = "";
}

public sealed record TransferInitRequest
{
    public string FileName { get; init; } = "";
    public long FileSize { get; init; }
    public string? MimeType { get; init; }
    public string Sha256 { get; init; } = "";
    public int ChunkSize { get; init; }
}

public sealed record TransferInitResponse
{
    public string TransferId { get; init; } = "";
    public bool Accepted { get; init; }
    public List<int> MissingChunks { get; init; } = new();
}

public sealed record TransferChunkRequest
{
    public string TransferId { get; init; } = "";
    public int Index { get; init; }
    public int TotalChunks { get; init; }
    public string Nonce { get; init; } = "";
    public string CipherText { get; init; } = "";
    public string Aad { get; init; } = "";
}

public sealed record TransferChunkResponse
{
    public bool Received { get; init; }
}

public sealed record TextMessageRequest
{
    public string MessageId { get; init; } = "";
    public string Nonce { get; init; } = "";
    public string CipherText { get; init; } = "";
    public string Aad { get; init; } = "";
}

public sealed record TextMessageResponse
{
    public bool Received { get; init; }
}






